### PR TITLE
refactor(postgres): Unit test SaveEvent

### PR DESCRIPTION
**Pull request by bndw**
Refactors SaveEvent so it's unit testable and adds tests to assert the current behavior.

### CODE REVIEW

File: storage/postgresql/save.go [hunk 1]
```diff
@@ -9,30 +9,13 @@ import (
...
-	} else if evt.Kind >= 30000 && evt.Kind < 40000 {
-		// NIP-33
-		d := evt.Tags.GetFirst([]string{"d"})
-		if d != nil {
-			b.DB.Exec(`DELETE FROM event WHERE pubkey = $1 AND kind = $2 AND tagvalues && ARRAY[$3]`,
-				evt.PubKey, evt.Kind, d.Value())
-		}
+	deleteQuery, deleteParams, shouldDelete := deleteBeforeSaveSql(evt)
+	if shouldDelete {
+		_, _ = b.DB.ExecContext(ctx, deleteQuery, deleteParams...)
```
Comment (bndw) [inline storage/postgresql/save.go:14]: There's opportunity for error handling here but I didn't want to change any of the existing behavior without double checking.

File: storage/postgresql/save.go [hunk 2]
```diff
@@ -60,3 +43,47 @@ func (b *PostgresBackend) AfterSave(evt *nostr.Event) {
...
+}
+
+func saveEventSql(evt *nostr.Event) (string, []any, error) {
+	const query = `INSERT INTO event (
+	id, pubkey, created_at, kind, tags, content, sig)
+	VALUES ($1, $2, $3, $4, $5, $6, $7)
+	ON CONFLICT (id) DO NOTHING`
+
+	var (
+		tagsj, _ = json.Marshal(evt.Tags)
```
Comment (bndw) [inline storage/postgresql/save.go:84]: There's opportunity for error handling here but I didn't want to change any of the existing behavior without double checking.

**Comment by bndw** [comment]
[NOT URGENT] My motivation in unit testing the generated SQL is to give developers confidence they're not breaking existing behavior when making changes. 

For example I previously wanted to look at query performance but due to the size/complexity of QueryEvents I wasn't able to hold it in my head or feel comfortable changing anything, for fear of breaking an edge case. By adding unit tests that assert the generated SQL we can confidently make changes to query generation.

Let me know if you find this approach valuable or helpful; my intention was to add similar tests to all of the postgres SQL generation. I appreciate a lot of this is personal preference and this is your project - I respect your decision either way :)

**Comment by fiatjaf** [comment]
Sorry for not responding. I didn't do anything here because I was about to merge it, but something happened and I didn't, so I forgot. I guess I will merge now if you tell me this is ok to merge. It looks good to me.

**Comment by bndw** [comment]
It's no problem. I imagine you're quite busy these days :). It's good to merge.
